Abandoned: binutils: add support for version 2.39

hardfalcon opened the pull request:
Original commit cc24c4ed by Nick Hainke <vincent@systemli.org>

Changes in binutils 2.39:

* The ELF linker will now generate a warning message if the stack is made executable.  Similarly it will warn if the output binary contains a segment with all three of the read, write and execute permission bits set.  These warnings are intended to help developers identify programs which might be vulnerable to attack via these executable memory regions. The warnings are enabled by default but can be disabled via a command line option.  It is also possible to build a linker with the warnings disabled, should that be necessary.
* The ELF linker now supports a --package-metadata option that allows embedding a JSON payload in accordance to the Package Metadata specification.
* In linker scripts it is now possible to use TYPE=<type> in an output section description to set the section type value.
* The objdump program now supports coloured/colored syntax highlighting of its disassembler output for some architectures. (Currently: AVR, RiscV, s390, x86, x86_64).
* The nm program now supports a --no-weak/-W option to make it ignore weak symbols.
* The readelf and objdump programs now support a -wE option to prevent them from attempting to access debuginfod servers when following links.
* The objcopy program's --weaken, --weaken-symbol, and --weaken-symbols options now works with unique symbols as well.

Announcement:
https://sourceware.org/pipermail/binutils/2022-August/122246.html

Added upstream commit fe39ffdc to fix building issues with musl.

Fixes: cc24c4ed5eae66c061e5787bb172fd2a91861f25
Fixes: a9eb5cce7e5c4996fa6ffc081b5323bfd4735dc2

Signed-off-by: Nick Hainke <vincent@systemli.org>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

PolynomialDivision committed:
binutils: update to 2.39

Changes:

* The ELF linker will now generate a warning message if the stack is made
  executable.  Similarly it will warn if the output binary contains a
  segment with all three of the read, write and execute permission
  bits set.  These warnings are intended to help developers identify
  programs which might be vulnerable to attack via these executable
  memory regions.
  The warnings are enabled by default but can be disabled via a command
  line option.  It is also possible to build a linker with the warnings
  disabled, should that be necessary.
* The ELF linker now supports a --package-metadata option that allows
  embedding a JSON payload in accordance to the Package Metadata
  specification.
* In linker scripts it is now possible to use TYPE=<type> in an output
  section description to set the section type value.
* The objdump program now supports coloured/colored syntax
  highlighting of its disassembler output for some architectures.
  (Currently: AVR, RiscV, s390, x86, x86_64).
* The nm program now supports a --no-weak/-W option to make it ignore
  weak symbols.
* The readelf and objdump programs now support a -wE option to prevent
  them from attempting to access debuginfod servers when following
  links.
* The objcopy program's --weaken, --weaken-symbol, and
  --weaken-symbols options now works with unique symbols as well.

Announcement:
https://sourceware.org/pipermail/binutils/2022-August/122246.html

Signed-off-by: Nick Hainke <vincent@systemli.org>

Ansuel commented:
@hardfalcon Hi, thanks for the patch, you can't remove the sob tag if the commit is based on that with minor changes.... 

I would add his sob and your on top of it.

hardfalcon commented:
@Ansual: I was reluctant keeping @PolynomialDivision's `Signed-off-by:` line in because his original version didn't include the patch file from upstream and thus I didn't want to give the impression as if he had signed off on including that patch file (and I wasn't sure if a patch file this big/substancial would be considered "minor"). But since you say this is fine, I've rebased my Pull Request and included the additional `Signed-off-by:` in both the pull request description and the rebased commit's commit message. :)

PolynomialDivision commented:
This is not a toolchain bump. ;) You update a openwrt package, so prefix is just `binutils: `. :)

hauke commented:
This does not compile for me. I get this error message:
````
make[7]: Entering directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng'
Making all in libcollector
make[8]: Entering directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng/libcollector'
make  all-am
make[9]: Entering directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng/libcollector'
/bin/bash ./libtool  --tag=CC   --mode=compile aarch64-openwrt-linux-musl-gcc -DHAVE_CONFIG_H -I.  -U_ASM -I.. -I. -I./../common -I./../src -I./../../include  -I../../bfd -I./../..  -Wall -Wno-nonnull-compare -O
s -pipe -mcpu=cortex-a53 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result   -fmacro-prefix-map=/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53
_musl/binutils-2.39=binutils-2.39 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fPIC -Wno-unused-value      -MT libgp_collector_la-libcol_hwcdrv.lo -MD -MP
 -MF .deps/libgp_collector_la-libcol_hwcdrv.Tpo -c -o libgp_collector_la-libcol_hwcdrv.lo `test -f 'libcol_hwcdrv.c' || echo './'`libcol_hwcdrv.c
OpenWrt-libtool-patched-2.2: compile:  aarch64-openwrt-linux-musl-gcc -DHAVE_CONFIG_H -I. -U_ASM -I.. -I. -I./../common -I./../src -I./../../include -I../../bfd -I./../.. -Wall -Wno-nonnull-compare -Os -pipe -mc
pu=cortex-a53 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -fmacro-prefix-map=/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutil
s-2.39=binutils-2.39 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fPIC -Wno-unused-value -MT libgp_collector_la-libcol_hwcdrv.lo -MD -MP -MF .deps/libgp_c
ollector_la-libcol_hwcdrv.Tpo -c libcol_hwcdrv.c  -fPIC -DPIC -o .libs/libgp_collector_la-libcol_hwcdrv.o
In file included from ./../common/hwcdrv.c:35,
                 from libcol_hwcdrv.c:25:
./../common/cpuid.c:27:1: error: expected '=', ',', ';', 'asm' or '__attribute__' before '__get_cpuid'
   27 | __get_cpuid (unsigned int op ATTRIBUTE_UNUSED, unsigned int *eax,
      | ^~~~~~~~~~~
make[9]: *** [Makefile:695: libgp_collector_la-libcol_hwcdrv.lo] Error 1
make[9]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng/libcollector'
make[8]: *** [Makefile:479: all] Error 2
make[8]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng/libcollector'
make[7]: *** [Makefile:471: all-recursive] Error 1
make[7]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng'
make[6]: *** [Makefile:403: all] Error 2
make[6]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39/gprofng'
make[5]: *** [Makefile:7414: all-gprofng] Error 2
make[5]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39'
make[4]: *** [Makefile:1004: all] Error 2
make[4]: Leaving directory '/home/hauke/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/binutils-2.39'
````

I used this configuration:
````
hauke@server-dev:~/openwrt/openwrt$ ./scripts/diffconfig.sh 
CONFIG_TARGET_armvirt=y
CONFIG_TARGET_armvirt_64=y
CONFIG_TARGET_armvirt_64_Default=y
CONFIG_ALL=y
CONFIG_ALL_KMODS=y
CONFIG_ALL_NONSHARED=y
# CONFIG_PACKAGE_unet-cli is not set
# CONFIG_PACKAGE_unetd is not set
````

hauke commented:
Please rebase to trigger a run of the new "Build all core packages" workflow.

hardfalcon commented:
@PolynomialDivision: Thanks for the hint, I've fixed the commit message. :)

@hauke: Sorry for not responding earlier, I've rebased the commit now.

hauke commented:
There are still two build problems in this change.
bpftools does not build on MIPS malta:
```
====== Make errors from logs/package/network/utils/bpftools/full/compile.txt ======
/__w/openwrt/openwrt/openwrt/staging_dir/target-mips-openwrt-linux-musl_musl/usr/include/dis-asm.h:472:13: note: declared here
  472 | extern void init_disassemble_info (struct disassemble_info *dinfo, void *stream,
      |             ^~~~~~~~~~~~~~~~~~~~~
jit_disasm.c:105:17: error: too few arguments to function 'init_disassemble_info'
  105 |                 init_disassemble_info(&info, stdout,
      |                 ^~~~~~~~~~~~~~~~~~~~~
In file included from jit_disasm.c:23:
/__w/openwrt/openwrt/openwrt/staging_dir/target-mips-openwrt-linux-musl_musl/usr/include/dis-asm.h:472:13: note: declared here
  472 | extern void init_disassemble_info (struct disassemble_info *dinfo, void *stream,
      |             ^~~~~~~~~~~~~~~~~~~~~
make[4]: *** [Makefile:216: /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/bpftools-full/bpftools-2022-03-08-04c465fd/jit_disasm.o] Error 1
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/bpftools-full/bpftools-2022-03-08-04c465fd/src'
make[3]: *** [Makefile:150: /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/bpftools-full/bpftools-2022-03-08-04c465fd/.built] Error 2
time: package/network/utils/bpftools/full/compile#3.34#0.54#3.91
Error: Process completed with exit code 2.
```
https://github.com/openwrt/openwrt/actions/runs/3360783934/jobs/5570395757

binutils does not build on x86/64:
```
====== Make errors from logs/package/devel/binutils/compile.txt ======
rstrip.sh: /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/objdump/usr/bin/objdump: removing rpath /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/bfd/.libs
Packaged contents of /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/objdump into /__w/openwrt/openwrt/openwrt/bin/packages/x86_64/base/objdump_2.39-1_x86_64.ipk
rstrip.sh: /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/ar/usr/bin/ar: executable
rstrip.sh: /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/ar/usr/bin/ar: removing rpath /__w/openwrt/openwrt/openwrt/openwrt-toolchain-x86-64_gcc-11.3.0_musl.Linux-x86_64/toolchain-x86_64_gcc-11.3.0_musl/bin//../../lib
rstrip.sh: /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/ar/usr/bin/ar: removing rpath /__w/openwrt/openwrt/openwrt/openwrt-toolchain-x86-64_gcc-11.3.0_musl.Linux-x86_64/toolchain-x86_64_gcc-11.3.0_musl/bin//../../usr/lib
rstrip.sh: /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/ar/usr/bin/ar: removing rpath /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/bfd/.libs
Packaged contents of /__w/openwrt/openwrt/openwrt/build_dir/target-x86_64-openwrt-linux-musl_musl/binutils-2.39/ipkg-x86_64/ar into /__w/openwrt/openwrt/openwrt/bin/packages/x86_64/base/ar_2.39-1_x86_64.ipk
Package binutils is missing dependencies for the following libraries:
libgprofng.so.0
libstdc++.so.6
make[3]: *** [Makefile:153: /__w/openwrt/openwrt/openwrt/bin/packages/x86_64/base/binutils_2.39-1_x86_64.ipk] Error 1
time: package/devel/binutils/compile#304.23#42.19#359.70
Error: Process completed with exit code 2.
```
https://github.com/openwrt/openwrt/actions/runs/3360783934/jobs/5570395871

hardfalcon commented:
You're right. I had actually added this patch to both `toolchain/binutils/patches/2.39` and `package/devel/binutils/patches`, but I had hoped just adding it to `package/devel/binutils/patches` would be sufficient. I've now changed my commit such that the patch is moved from `package/devel/binutils/patches` to `toolchain/binutils/patches/2.39`.

hardfalcon commented:
Retrying with the patch in both directories since the build run failed with the patch being present in only one of the directories.

hardfalcon committed:
toolchain/binutils: Fix binutils 2.39 gprofng related build issues on musl

Patch taken from upstream commit fe39ffdc.
Check https://sourceware.org/PR29477 for details.

hardfalcon committed:
binutils: Fix binutils 2.39 gprofng related build issues on musl

Patch taken from upstream commit fe39ffdc.
Check https://sourceware.org/PR29477 for details.

hardfalcon committed:
toolchain/binutils: switch to version 2.39 by default

gkourachanis commented:
Hi, I don't know if this is the correct place to say this, but with toolchain-x86_64_gcc-12.2.0_musl (no tweaks, compiled with the default configuration) and the upstream patch from fe39ffdc202f04397f31557f17170b40bc42b77a (the one @hardfalcon has added in this WIP) the binutils package on version 2.39 compiles fine for me. The compilation runs fine on glibc as well (here also, gcc v12.2.0).

If there's anything I can do to provide info, tell me what to do.

P.S.: I was about to pull request the patch, and I just found out this thread and that it was already being reviewed. I should be reading more of the issues in the future :p
P.S.: I haven't updated my forked openwrt repository with the version that I mention that succeeds. I will probably do it soon. Not that it has any major patches anyway.

hauke commented:
The github CI still sees a compile problem in the `bpftools` package

hardfalcon commented:
Sorry, I'm currently not getting around to dig into this further, and clearly my past tries have all failed. I'll close this one, feel free to pick it up if you want to.

If you do, I really hope you will be more successful than I was. No need to credit me with anything, I merely took a single commit from upstream and noticed that it works for me. :)
